src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,812 @@
+# Table of Contents
+
+- [Terminology](#terminology)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+  - [In depth](#in-depth)
+  - [Compatibility](#compatibility)
+- [Examples](#examples)
+  - [Using Default on Serialize](#using-default-on-serialize)
+  - [Using Ignore on Serialize](#using-ignore-on-serialize)
+  - [Using Preserve on Serialize](#using-preserve-on-serialize)
+  - [Using Preserve on Deserialize](#using-preserve-on-deserialize)
+- [Other languages](#other-languages)
+  - [Newtonsoft.Json](#newtonsoftjson)
+  - [dojo toolkit (JavaScript framework)](#dojo-toolkit-javascript-framework)
+  - [flatted (JavaScript module) (probably not worth it)](#flatted-javascript-module-probably-not-worth-it)
+  - [Jackson (Java)](#jackson-java)
+  - [golang](#golang)
+- [Ground rules](#ground-rules)
+  - [Reference objects ($ref)](#reference-objects-ref)
+  - [Preserved objects ($id)](#preserved-objects-id)
+  - [Preserved arrays](#preserved-arrays)
+  - [JSON Objects if not Collection (Class | Struct | Dictionary) - On Deserialize (and Serialize?)](#json-objects-if-not-collection-class--struct--dictionary---on-deserialize-and-serialize)
+  - [JSON Object if Collection - On Deserialize](#json-object-if-collection---on-deserialize)
+  - [Immutable types](#immutable-types)
+  - [Value types](#value-types)
+  - [Interaction with JsonPropertyNameAttribute](#interaction-with-jsonpropertynameattribute)
+- [Future](#future)
+- [Notes](#notes)
+
+
+# Terminology
+
+**Reference loops**: Also referred as circular references, loops occur when a property of a .NET object refers to the object itself, either directly (a -> a) or indirectly (a -> b -> a). They also occur when the element of an array refers to the array itself (arr[0] -> arr). Multiple occurrences of the same reference do not imply a cycle.
+
+**Preserve duplicated references**: Semantically represent objects and/or arrays that have been previously written, with a reference to them when found again in the object graph (using reference equality for comparison).
+
+**Metadata**: Extra properties on JSON objects and/or arrays (that may change their schema) to enable reference preservation when round-tripping. These additional properties are only meant to be understood by the `JsonSerializer`.  
+
+# Motivation
+
+Currently, there is no mechanism to avoid infinite loops while serializing .NET object instances that contain cycles nor to preserve references that round-trip when using `System.Text.Json`. The `JsonSerializer` throws a `JsonException` when a loop is found within the object graph.
+
+This is a heavily requested feature since it is considered by many as a very common scenario, specially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).
```

Comment:
```suggestion
This is a heavily requested feature since it is considered by many as a very common scenario, especially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).
```